tests/check_prog.py

```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python3
+import argparse
+import pathlib
+import os
+import subprocess
+import logging as log
+
+SEED = 3370029442
+
+FILE_DIR = pathlib.Path(__file__).parent.resolve()
+REFERENCE_DIR = FILE_DIR.joinpath("references/")
+
+
+EXIT_SUCCESS = 0
+EXIT_FAILURE = 1
+
+
+def exec_process(cmd, *args, silent=False, **kwargs):
+    log.debug("Executing %s ", cmd)
+    result = subprocess.run(cmd.split(), stdout=subprocess.PIPE,
+                            stderr=subprocess.PIPE, *args, **kwargs)
+    if result.stdout:
+        log.debug("Process output: %s", result.stdout.decode("utf-8"))
+    if result.returncode != EXIT_SUCCESS and not silent:
+        log.error("BEGIN %s", 40 * "#")
+        log.error("Failed while executing:\n%s\n", cmd)
+        log.error("Output:\n%s", result.stderr.decode("utf-8"))
+        log.error("END %s", 40 * "#")
+
+    return result
+
+
+def main(args):
+
+    COMPILER_BIN = args.compiler.absolute()
+    VALIDATION_BIN = args.validation.absolute()
+    PRUNER_BIN = args.pruner_path.absolute()
+    P4_PROG = args.p4prog.absolute()
+
+    if not os.access(COMPILER_BIN, os.X_OK):
```

Comment:
Use pathlib.exists instead.

Reply:
Also a nice trick is to use a lambda in arg parse. For example:

```
  parser.add_argument("-i", "--p4_input", dest="p4_input", default=None,
                      type=lambda x: util.is_valid_file(parser, x),
                      help="The main reference p4 file.")
```

Reply:
I wanted to check if the provided file has the executable bit set.

Reply:
Oh I see, there is no builtin facility for this but I think shutil.which does the job well enough.
Also I would convert all prints to log statements